### DISCUSSION TITLE
Add dtype_backend to fetchdf()

### DISCUSSION THREAD

**Discussion by Dino-Burger** [duckdb/duckdb#8814] [Ideas]
Hi All

I often use duckdb to emit pandas dataframes. So far all dataframes generated by duckdb have the normal pandas datatypes (to my best knowledge). It would be great if we could add a parameter `dtype_backend` like the one supported in pandas.read_* functions (like [read_csv](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)) to change the underlying datatypes to pyarrow datatypes.

All elements to achieve this are present in my opinion, so below is some code to monkey patch this functionality:

```python
def enable_dtype_backend(duckdb_module):
    import pandas as pd
    duckdb_module.DuckDBPyRelation.original_fetchdf = duckdb_module.DuckDBPyRelation.fetchdf
    def df(self: duckdb_module.DuckDBPyRelation, date_as_object: bool = False, dtype_backend: str = "numpy_nullable"):
        if dtype_backend == "numpy_nullable":
            return self.original_fetchdf(date_as_object=date_as_object)
        elif dtype_backend == "pyarrow":
            return self.arrow().to_pandas(date_as_object=date_as_object, types_mapper=pd.ArrowDtype)
    duckdb_module.DuckDBPyRelation.fetchdf = df
    duckdb_module.DuckDBPyRelation.df = df

import duckdb
enable_dtype_backend(duckdb)

duckdb.query("Select NULL::int64 as a").df().dtypes
# returns: a    float64
# This conversion on ints to floats is what I wish to avoid

duckdb.query("Select NULL::int64 as a").df(dtype_backend = "pyarrow").dtypes
# returns: a    int64[pyarrow]
```

Thanks to all working on the fabulous duckdb!

Dino

**Comment by Tishj** [comment]
Thanks, I've had this same idea before
The one thing stopping me is that the pandas arrow dtype is still experimental

**Reply by Dino-Burger** [reply]
Cool thanks!
Was also thinking about it and I think having to pass this argument every time is not so great. So let’s hope pandas will soon label the arrow backend stable and define a general switch for all dataframes in a program.

Good day
